README.md
# Rowing Machine Interface

This is basic module to standardize outputs from rowing machines (via either USB or Bluetooth connections).

As of this current release, it only supports the [WaterRower](https://www.waterrower.com/us/) (with the [S4 module](https://www.waterrower.com/us/shop/accessories/commodule.html)).

This is intended to power apps such as the [Home Rower Game](https://github.com/ecaron/home-rower-game), but minimize the time those apps have to spend handling the nuances of different rowing devices.

## Sample Code

Although more complicated examples are in the [demo](demo) directory, a basic application could work like this:

```javascript
const rower = require('rowing-machine-interface')

const run = async function () {
  let rower

  try {
    rower = await app({})
    console.log('Connected, awaiting start')
  } catch (e) {
    connectionStatus(chalk.red('Connection failed'))
    return
  }
  rower.event.on('state', (state) => {
    console.log(`Received state: ${JSON.stringify(state)}`)
  })
  rower.event.on('update', (memory) => {
    console.log(`Received update: ${JSON.stringify(memory)}`)
  })
}

run()
```

## Event Emitted

The module emits one of two events, with additional information for each:
* `state`
  * `status` (either `started`, `error`, `reconnecting`, `reset`, `stopped`)
  * `details`
* `update`
  * `memory` (The metric that was modifiied)

## Metrics Tracked
* Distance in meters (as `ms_distance`)
* Average time for a pull (as `stroke_pull`)
* Average time for a whole stroke (as `stroke_average`)
* Stroke Count (as `stroke_cnt`)
* Speed in knots (as `ms_average`)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0]

Initial release

### Added
- Connection to S4 Water Rower, with just 2 common package dependencies
- Streamlining all the events emitted
- Added ability to have counters go to zero on reset (or not)
- Using streams, information from the rower can be sent to external systems (such as database)
- Include simple demo to record & playback a rowing session

s4/index.js
const Rower = require('./rower')
const RowerPlayback = require('./playback')
const memoryMap = require('./memory-map')
const debug = require('debug')('S4')
let rower

exports.rower = async function (config) {
  if (rower && !rower.destroyed) {
    return rower
  }

  if (config.record && typeof config.record !== 'object') {
    throw new Error('`record` must be a stream`')
  }

  rower = new Rower(config, memoryMap)

  config.port = false
  try {
    config.port = await rower.findPort()
  } catch (e) {
    throw new Error('Rowing machine port not found')
  }

  if (config.port !== false) {
    debug('[Init] Port is available - starting rower')
    rower.start()
    return rower
  }
  throw new Error('Rowing machine not found')
}

exports.playback = function (config) {
  if (!config.playback) {
    throw new Error('Playback stream not specified')
  }
  const rower = new RowerPlayback(config, memoryMap)
  return rower
}


s4/rower.js
const { EventEmitter } = require('events')
const { SerialPort } = require('serialport')
const { ReadlineParser } = require('@serialport/parser-readline')
const debug = require('debug')('S4:rower')

class S4 {
  constructor (config, memoryMap) {
    this.config = config
    this.memoryMap = memoryMap
    this.event = new EventEmitter()
    this.port = null
    this.pending = []
    this.writer = null
    this.connected = false
    this.reconnecting = false
    this.recording = false
    this.destroyed = false
    this.timer = new Date()
    if (config.record) {
      this.recording = config.record
    }

    this.EOL = '\r\n' // CRLF 0x0D0A
  }

  async findPort () {
    const ports = await SerialPort.list()
    for (let i = 0; i < ports.length; i++) {
      if (typeof ports[i].vendorId !== 'string' || typeof ports[i].productId !== 'string') {
        continue
      }
      // https://usb-ids.gowdy.us/read/UD/04d8/000a
      if (ports[i].vendorId.toLowerCase() === '04d8' && ports[i].productId.toLowerCase() === '000a') {
        // port is an object literal with string values
        return ports[i].path
      }
    }
    debug('USB device not detected')
    return false
  }

  start () {
    if (!this.config.port) {
      return false
    }
    const self = this
    this.port = new SerialPort({ path: this.config.port, baudRate: 19600, autoOpen: false, lock: false })
    const parser = this.port.pipe(new ReadlineParser({ delimiter: '\r\n' }))
    parser.on('data', this._readAndDispatch.bind(this))
    this._open()
    this.port.on('error', function (err) {
      debug('Port errored, trying to open again in 5 seconds')
      debug(err)
      self._reconnect('Port errored, trying to open again in 5 seconds')
    })
    this.port.on('close', function () {
      debug('Port was closed, trying again in 5 seconds')
      self._reconnect('Port was closed, trying again in 5 seconds')
    })
  }

  getMemory () {
    return this.memoryMap
  }

  reset () {
    this.pending = []
    this._write('RESET')
    this._emit('state', { status: 'reset' })
    if (this.config.zeroOnReset) {
      this.memoryMap.forEach(response => {
        response.value = 0
      })
    }
  }

  destroy () {
    this._emit('state', { status: 'stopped' })
    this._write('EXIT')
    this.pending = []
    this.destroyed = true
    if (this.writer) {
      clearInterval(this.writer)
    }
    if (this.recording) {
      this.recording.end()
    }
  }

  _emit (event, details) {
    this.event.emit(event, details)
    if (this.recording) {
      const now = new Date()
      const sinceLast = now - this.timer
      this.timer = now
      this.recording.write(`["${event}",${sinceLast},${JSON.stringify(details)}]\n`)
    }
  }

  _write (string) {
    this.pending.push(string)
  }

  _flushNext () {
    if (this.pending.length === 0) {
      return
    }

    const string = this.pending.shift()
    if (this.port) {
      const buffer = Buffer.from(string + this.EOL)
      // debug('[OUT]: ' + buffer)
      this.port.write(buffer)
      if (string === 'RESET') {
        this._emit('state', { status: 'started' })
      }
    } else {
      debug(`Communication port is not open - not sending data: ${string}`)
      this._emit('state', {
        status: 'error',
        details: `Communication port is not open - not sending data: ${string}`
      })
    }
  }

  _readAndDispatch (string) {
    // debug('[IN]: ' + string)
    const c = string.charAt(0)
    switch (c) {
      case '_':
        this._wrHandler(string)
        break
      case 'I':
        this._informationHandler(string)
        break
      case 'O':
        // ignore
        break
      case 'E':
        // ignore
        break
      case 'P':
        // ignore
        break
      case 'S':
        for (let i = 0; i < this.memoryMap.length; i++) {
          this._readMemoryAddress(this.memoryMap[i].address, this.memoryMap[i].size)
        }
        break
      default:
        this._unknownHandler(string)
    }
  }

  // handlers start
  _unknownHandler (string) {
    debug(`Unrecognized packet: ${string}`)
  }

  _wrHandler (string) {
    if (string === '_WR_') {
      this._write('IV?')
    } else {
      this._unknownHandler(string)
    }
  }

  _informationHandler (string) {
    const c = string.charAt(1)
    switch (c) {
      case 'V':
        this._informationVersionHandler(string)
        break
      case 'D':
        this._memoryValueHandler(string)
        break
      default:
        this._unknownHandler(string)
    }
  }

  _readMemoryAddress (address, size) {
    const cmd = 'IR' + size + address
    this._write(cmd)
  }

  _informationVersionHandler (string) {
    // IV40210
    const model = string.charAt(2)
    const fwRevMajor = string.substring(3, 5)
    const fwRevMinor = string.substring(5, 7)
    const version = 'S' + model + ' ' + fwRevMajor + '.' + fwRevMinor
    // only log error, ignore version mismatch
    if (version !== 'S4 02.10') {
      debug(`WaterRower monitor version mismatch - expected S4 02.10 but got ${version}`)
      this._emit('state', {
        status: 'error',
        details: `WaterRower monitor version mismatch - expected S4 02.10 but got ${version}`
      })
    } else {
      debug('WaterRower ' + version)
    }
    this.reset()
  }

  _memoryValueHandler (string) {
    const size = string.charAt(2)
    const address = string.substring(3, 6)
    let l
    switch (size) {
      case 'S':
        l = 1
        break
      case 'D':
        l = 2
        break
      case 'T':
        l = 3
        break
      default:
        this._unknownHandler(string)
        return
    }
    const end = 6 + 2 * l
    const dataPoint = this.memoryMap.find(element => element.address === address)
    let value = parseInt(string.substring(6, end), 16)
    if (dataPoint.multiple) {
      value = value * dataPoint.multiple
    }
    if (dataPoint.value !== value) {
      debug(`${dataPoint.name} changed from ${dataPoint.value} to ${value}`)
      dataPoint.prevValue = dataPoint.value
      dataPoint.value = value
      this._emit('update', dataPoint)
    }
    this._readMemoryAddress(address, size)
  }

  toString () {
    return this.port.path
  }

  _reconnect (message) {
    if (!message) {
      message = ''
    } else {
      message += '  '
    }
    this.connected = false
    if (this.reconnecting) {
      debug('Already trying to reconnect')
      return
    }
    this._emit('state', {
      status: 'reconnecting',
      details: message + 'Attempting reconnect in 5s'
    })
    debug('Attempting reconnect in 5s')
    this.reconnecting = setTimeout(this._open.bind(this), 5000)
  }

  _open () {
    const self = this
    clearTimeout(this.reconnecting)
    this.reconnecting = false
    this.port.open(function (err) {
      if (err) {
        debug('Error occorred opening connection. Attempting reconnect.')
        self._reconnect('Error occorred opening connection. Attempting reconnect.')
        return
      }

      // we can only write one message every .2s
      if (self.writer) {
        clearInterval(self.writer)
        if (self.pending.length > 0) {
          self.reset()
        }
      }
      self.writer = setInterval(self._flushNext.bind(self), 200)

      debug('Connected successfully')
      self.connected = true
      self._write('USB')
    })
  }
}

module.exports = S4


demo/player.js
const fs = require('fs')
const readline = require('readline')
const chalk = require('chalk')
const app = require('../app')
require('draftlog').into(console)

if (process.argv.length !== 3) {
  console.warn('Script must be called `node playback [recording-file-name]')
  process.exit()
}

const connectionStatus = console.draft(chalk.yellow(`Playing back recording "${process.argv[2]}"`))

const run = async function () {
  let rower
  try {
    rower = await app({ playback: true })
    connectionStatus(chalk.green('Loading & starting'))
    rower.start()
  } catch (e) {
    connectionStatus(chalk.red('Load failed'))
    return
  }

  const readInterface = readline.createInterface({
    input: fs.createReadStream(process.argv[2])
  })
  readInterface.on('line', function (line) {
    rower.load(JSON.parse(line))
  })

  rower.event.on('state', (state) => {
    if (state.status === 'started') {
      connectionStatus(chalk.green('Connection started'))
    } else if (state.status === 'stopped') {
      connectionStatus(chalk.red('Connection stopped'))
    } else if (state.status === 'reconnecting') {
      connectionStatus(chalk.yellow('Connection stopped'))
    } else if (state.status === 'reset') {
      connectionStatus(chalk.yellow('Connection reset'))
    } else {
      connectionStatus(chalk.error(`Error: ${state.details}`))
    }
  })
  const statusLines = {}
  rower.getMemory().forEach(memory => {
    statusLines[memory.name] = {
      message: `${('[' + memory.name.padEnd(14, ' ') + ']').padEnd(47, ' ')}${' - ' + memory.desc}`,
      timer: false
    }
    statusLines[memory.name].draft = console.draft(chalk.dim(statusLines[memory.name].message))
  })
  rower.event.on('update', (memory) => {
    let lineEntry = `[${memory.name.padEnd(14, ' ')}] - `
    let memoryValue = memory.value
    if (typeof memoryValue === 'number' && Math.floor(memoryValue) !== memoryValue) {
      memoryValue = memoryValue.toFixed(3)
    }
    if (memory.prevValue) {
      let prevValue = memory.value
      if (typeof prevValue === 'number' && Math.floor(prevValue) !== prevValue) {
        prevValue = prevValue.toFixed(3)
      }
      lineEntry += `${memoryValue + ' ' + memory.unit} (was ${prevValue}) `
    } else {
      lineEntry += `${memoryValue + ' ' + memory.unit} (first data) `
    }
    clearTimeout(statusLines[memory.name].timer)
    statusLines[memory.name].message = `${lineEntry.padEnd(47, ' ')}${chalk.gray(' - ' + memory.desc)}`
    statusLines[memory.name].draft(statusLines[memory.name].message)
    statusLines[memory.name].timer = setTimeout(function () {
      statusLines[memory.name].draft(chalk.dim(statusLines[memory.name].message))
    }, 5000)
  })
}

run()


s4/playback.js
const { EventEmitter } = require('events')
const debug = require('debug')('S4:playback')

class S4 {
  constructor (config, memoryMap) {
    this.config = config
    this.memoryMap = memoryMap
    this.event = new EventEmitter()
    this.timer = null
    this.data = []
  }

  load (entry) {
    this.data.push(entry)
  }

  start () {
    let part
    if (this.data.length > 0) {
      part = this.data.shift()
      this.event.emit(part[0], part[2])
      if (part[0] === 'update') {
        debug(`${part[2].name} changed from ${part[2].prevValue} to ${part[2].value}`)
        const dataPoint = this.memoryMap.find(element => element.address === part[2].address)
        dataPoint.prevValue = part[2].prevValue
        dataPoint.value = part[2].value
      }
    }
    if (this.data.length > 0) {
      this.timer = setTimeout(this.start.bind(this), part[1])
    } else {
      // Wait for more data, or a destroy to be called
      this.timer = setTimeout(this.start.bind(this), 1000)
    }
  }

  getMemory () {
    return this.memoryMap
  }

  destroy () {
    if (this.timer) {
      clearTimeout(this.timer)
    }
  }
}

module.exports = S4


demo/listener.js
const fs = require('fs')
const chalk = require('chalk')
const app = require('../app')
require('draftlog').into(console)
const connectionStatus = console.draft(chalk.yellow('Attempting connection to rower'))

const run = async function () {
  let rower
  const config = {}

  if (process.argv.length === 3) {
    config.record = fs.createWriteStream(process.argv[2], { flags: 'wx' })
  }

  try {
    rower = await app(config)
    connectionStatus(chalk.green('Connected, awaiting start'))
  } catch (e) {
    connectionStatus(chalk.red('Connection failed'))
    return
  }
  rower.event.on('state', (state) => {
    if (state.status === 'started') {
      connectionStatus(chalk.green('Connection started'))
    } else if (state.status === 'stopped') {
      connectionStatus(chalk.red('Connection stopped'))
    } else if (state.status === 'reconnecting') {
      connectionStatus(chalk.yellow('Connection stopped'))
    } else if (state.status === 'reset') {
      connectionStatus(chalk.yellow('Connection reset'))
    } else {
      connectionStatus(chalk.error(`Error: ${state.details}`))
    }
  })
  const statusLines = {}
  rower.getMemory().forEach(memory => {
    statusLines[memory.name] = {
      message: `${('[' + memory.name.padEnd(14, ' ') + ']').padEnd(47, ' ')}${' - ' + memory.desc}`,
      timer: false
    }
    statusLines[memory.name].draft = console.draft(chalk.dim(statusLines[memory.name].message))
  })
  rower.event.on('update', (memory) => {
    let lineEntry = `[${memory.name.padEnd(14, ' ')}] - `
    let memoryValue = memory.value
    if (typeof memoryValue === 'number' && Math.floor(memoryValue) !== memoryValue) {
      memoryValue = memoryValue.toFixed(3)
    }
    if (memory.prevValue) {
      let prevValue = memory.value
      if (typeof prevValue === 'number' && Math.floor(prevValue) !== prevValue) {
        prevValue = prevValue.toFixed(3)
      }
      lineEntry += `${memoryValue + ' ' + memory.unit} (was ${prevValue}) `
    } else {
      lineEntry += `${memoryValue + ' ' + memory.unit} (first data) `
    }
    clearTimeout(statusLines[memory.name].timer)
    statusLines[memory.name].message = `${lineEntry.padEnd(47, ' ')}${chalk.gray(' - ' + memory.desc)}`
    statusLines[memory.name].draft(statusLines[memory.name].message)
    statusLines[memory.name].timer = setTimeout(function () {
      statusLines[memory.name].draft(chalk.dim(statusLines[memory.name].message))
    }, 5000)
  })
}

run()
